### PR TITLE
fix(statusline): also allow right click when 'mousemodel' is "popup*"

### PR DESCRIPTION
**Pull request by luukvbaal**
Problem:    The 'statusline'-format ui elements did not receive right
            click events when "mousemodel" is "popup*"
Solution:   Do not draw popupmenu and handle click event instead.

### CODE REVIEW

**Comment by luukvbaal** [comment]
Should this be `fix` and can we backport to `0.9`? I can look into adding a test tomorrow.

**Comment by luukvbaal** [comment]
> apparently this affects an existing test. I suppose it will have to change if we alter the behavior of statusline clicks with setpos mousemodel. Although I haven't looked into what the failing test is actually checking for yet.

So the existing `'mousemodel'` statusline click test seems to verify that clicking on the statusline extends the visual selection (by moving to the end of the buffer? IDK not sure what is the current expected behavior).

I would say we can safely change that behavior? `'statusline'`-element clicks should have precedence over the current behavior IMO.

**Comment by zeertzjq** [comment]
I think yes.

**Comment by luukvbaal** [comment]
Oh no it actually checks that the visual selection stays the same when right-clicking on statusline.
I guess we can keep the test and should make sure it passes instead then, while handling statusline clicks.

**Comment by luukvbaal** [comment]
> Oh no it actually checks that the visual selection stays the same when right-clicking on statusline.
I guess we can keep the test and should make sure it passes instead then, while handling statusline clicks.

This wasn't right either. I missed that the test actually cut the selection by executing the popupmenu callback (`<Down><CR>`). I updated the test accordingly to check that it (tries) to execute the statusline callback instead...

**Comment by luukvbaal** [comment]
I was running into issues again that made me abandon this last time. The problem is that we need to know whether a statusline element was clicked in `do_mouse()` but this detection logic is in `jump_to_mouse()`, which is not always executed, and shouldn't be when `'mousemodel'` is `"popup"`.

I think I was able to make it work properly this time without the general refactor I attempted last time but abandoned (in vim), by using the `MOUSE_SETPOS` flag to `jump_to_mouse()` instead...

**Comment by luukvbaal** [comment]
Hmm, perhaps we should verify that all statusline click tests now work with all `'mousemodel'`s.

**Comment by luukvbaal** [comment]
OK, I ended up adding the winbar/statuscolumn logic to `get_fpos_of_mouse()` instead of doing a `MOUSE_SETPOS` `jump_to_mouse()`. This way we can detect earlier if we don't want a popupmenu and handle statusline clicks instead without side-effects.

File: test/functional/ui/mouse_spec.lua [hunk 1]
```diff
@@ -1840,15 +1841,14 @@ describe('ui/mouse/input', function()
...
+    meths.set_option('statusline', '%@MyFunc@MyFunc%T')
     funcs.setreg('"', '')
     meths.win_set_cursor(0, {1, 9})
-    feed('vee')
     meths.input_mouse('right', 'press', '', 0, 5, 1)
     meths.input_mouse('right', 'release', '', 0, 5, 1)
-    feed('<Down><CR>')
+    eq('E117: Unknown function: MyFunc', eval('v:errmsg'))
     eq({1, 9}, meths.win_get_cursor(0))
-    eq('ran away', funcs.getreg('"'))
```
Comment (zeertzjq) [inline test/functional/ui/mouse_spec.lua:1852]: I don't think this block is still needed. Just remove this block.

File: src/nvim/mouse.c [hunk 2]
```diff
@@ -212,22 +212,30 @@ static int get_fpos_of_mouse(pos_T *mpos)
...
+  // compute the position in the buffer line from the posn on the screen
+  bool below_buffer = mouse_comp_pos(wp, &row, &col, &mpos->lnum);
+
+  if (!below_buffer && *wp->w_p_stc != NUL && mouse_col < win_col_off(wp)) {
+    return MOUSE_STATUSCOL;
+  }
 
   // winpos and height may change in win_enter()!
-  if (row + wp->w_winbar_height >= wp->w_height) {  // In (or below) status line
+  if (mouse_row + wp->w_winbar_height >= wp->w_height) {  // In (or below) status line
```
Comment (zeertzjq) [inline src/nvim/mouse.c:225]: Why is `mouse_row` used here to compare with window height? This doesn't make sense.

File: src/nvim/mouse.c [hunk 3]
```diff
@@ -212,22 +212,30 @@ static int get_fpos_of_mouse(pos_T *mpos)
...
-  if (col >= wp->w_width) {  // In vertical separator line
-    return IN_SEP_LINE;
+
+  if (winrow == -1 && wp->w_winbar_height != 0) {
+    return MOUSE_WINBAR;
   }
 
-  if (wp != curwin) {
-    return IN_UNKNOWN;
+  if (mouse_col >= wp->w_width) {  // In vertical separator line
```
Comment (zeertzjq) [inline src/nvim/mouse.c:233]: ditto

**Review by zeertzjq — CHANGES_REQUESTED** [review]

File: src/nvim/mouse.c (hunk 2 above)
Comment (luukvbaal) [reply src/nvim/mouse.c]: Mistake. I moved the `mouse_comp_pos()` call up to use its result in the statuscolumn comparison. After which the original `row/col` are no longer valid for the other comparisons.
Fixed now to use the intermediate `winrow/col`.

File: test/functional/ui/statuscolumn_spec.lua [hunk 4]
```diff
@@ -420,45 +422,46 @@ describe('statuscolumn', function()
...
-    eq('', eval("g:testvar"))
-    -- Check that statusline click doesn't register as statuscolumn click
-    meths.input_mouse('right', 'press', '', 0, 12, 0)
-    eq('', eval("g:testvar"))
-    -- Check that cmdline click doesn't register as statuscolumn click
-    meths.input_mouse('right', 'press', '', 0, 13, 0)
-    eq('', eval("g:testvar"))
-  end)
+  for _, model in ipairs(mousemodels) do
+    it("works with 'statuscolumn' clicks with mousemodel=" .. model, function()
```
Comment (zeertzjq) [inline test/functional/ui/statuscolumn_spec.lua:426]: 'mousemodel' is not set in this test

**Review by zeertzjq — APPROVED** [review]

**Pull request merged**